src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -104,18 +130,142 @@ public void run() {
             return;
         }
 
-        VerifyIotCertificate verifyIotCertificate = useCases.get(VerifyIotCertificate.class);
-        String[] certificateIds = registry.getAllCertificateIds();
+        Optional<ThingAssociations> associations = getThingsAssociatedWithCoreDevice();
+        associations.ifPresent(this::refresh);
+    }
+
+    /**
+     * Returns ThingAssociations which has information about what Things are associated with the core device and
+     * which things are no longer associated with the core device.
+     */
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private Optional<ThingAssociations> getThingsAssociatedWithCoreDevice() {
+        RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                .initialRetryInterval(Duration.ofSeconds(30)).maxRetryInterval(Duration.ofMinutes(3))
+                .maxAttempt(3)
+                .retryableExceptions(Arrays.asList(ThrottlingException.class, InternalServerException.class)).build();
+
+        try {
+            Stream<Thing> cloudThings = RetryUtils.runWithRetry(
+                    retryConfig, iotAuthClient::getThingsAssociatedWithCoreDevice,
+                            "get-things-associated-with-core-device", logger);
+
+            Stream<Thing> localThings = thingRegistry.getAllThings();
+            ThingAssociations associations =  ThingAssociations.create(cloudThings);
+            associations.setLocalThings(localThings);
+            return Optional.of(associations);
+        } catch (AccessDeniedException e) {
+            logger.atWarn().cause(e).log(
+                    "Access denied to list things associated with core device. Please make sure you "
+                    + "have setup the correct permissions.");
+        } catch (Exception e) {
+            logger.warn("Failed to get things associated to the core device. Retry will be schedule later");
+        }
+
+        return Optional.empty();
+    }
+
+    private void refresh(ThingAssociations associations) {
+        associations.getThings().forEach(this::refreshCertificateThingAttachment);
+        associations.getStaleThings().forEach(thingRegistry::deleteThing);
 
-        for (String certificateId : certificateIds) {
-            if (pemStore.exists(certificateId)) {
-                String pem = pemStore.getPem(certificateId).get();
-                verifyIotCertificate.apply(pem);
+        certificateRegistry.getAllCertificates()
+                .forEach(certificate -> {
+            if (associations.isAttachedToThing(certificate)) {
+                refreshCertificateValidity(certificate);
```

Comment:
should we also deleteCertificate() if a certificate is found invalid during `refreshCertificateValidity()` ?

(rest of the changes look good to me)

Reply:
We do that already as part of the useCase